### ISSUE TITLE
Always on top doesn't work for some reason.

### ISSUE THREAD

**Issue by Conusoid** [SnosMe/awakened-poe-trade#476]
After updating to 2.14.1 keeps appearing behind the game window (whether it's windowed or fullscreen windowed). Clicking on the window doesn't fix it.
Arch linux, i3wm, picom.

**Comment by dvastazion** [comment]
> Is there a way to fix this under kwin? I'd rather not use another window manager if I can avoid it.

downgrade to version 2.13.1 fix it for me:
`cd ~;rm -ef ~/.config/awakened-poe-trade/; wget https://github.com/SnosMe/awakened-poe-trade/releases/download/v2.13.1/Awakened-PoE-Trade-2.13.1.AppImage; chmod +x Awakened-PoE-Trade-2.13.1.AppImage; sudo chattr +i Awakened-PoE-Trade-2.13.1.AppImage; ./Awakened-PoE-Trade-2.13.1.AppImage`

**Comment by Cilantr0** [comment]
I know you posted the issue on linux but it's happening on windows os too but I think I found a temp solution for it. I disabled hardware acceleration in the settings tab and restarted. Now it works. So I hope that helps.

![hardware_acc_disabled](https://user-images.githubusercontent.com/10783824/139552015-13935e08-c46e-435f-afb4-c61618bb5dc5.png)

**Comment by ktzee** [comment]
> I know you posted the issue on linux but it's happening on windows os too but I think I found a temp solution for it. I disabled hardware acceleration in the settings tab and restarted. Now it works. So I hope that helps.
> 
> ![hardware_acc_disabled](https://user-images.githubusercontent.com/10783824/139552015-13935e08-c46e-435f-afb4-c61618bb5dc5.png)

yeah, this won't help on Linux. We need hw accel enabled for the window to be made transparent.

**Comment by Conusoid** [comment]
> We need hw accel enabled for the window to be made transparent.

It works fine w\o hw accel in i3wm, though I'm using picom compositor for the window to be transparent.

**Comment by ktzee** [comment]
> > We need hw accel enabled for the window to be made transparent.
> 
> It works fine w/o hw accel in i3wm, though I'm using picom compositor for the window to be transparent.

Did you configure picom some way specific? I use i3+picom too, and without hw accel, the overlay is solid black.

**Comment by Conusoid** [comment]
> Did you configure picom some way specific? I use i3+picom too, and without hw accel, the overlay is solid black.
Had same issue with `backend = 'glx'` in picom config (setting 'glx' solved screen tearing issue for me), so I've changed it back to default  `backend = 'xrender'`.

**Comment by ktzee** [comment]
Yeah, I just checked and even without hw accel, it's not solid black anymore. It definitely used to be. Something has changed in the way this is rendered.
However, even with hw accell off, initiating a price check (ctrl+D) does not bring the window in the foreground. I still need to manually open the overlay. Ctrl+Alt+D, instead, brings it to the foreground.

**Comment by cf8** [comment]
How about something like "open in browser" hotkey, i think that feature would be nice.
Should work the same as clicking on "Open in browser" button.

**Comment by SnosMe** [comment]
@cf8 no, https://github.com/SnosMe/awakened-poe-trade/issues/116#issuecomment-702395635

**Comment by eclipseo** [comment]
@thorsten-passfeld 

I had the same issue with KDE Plasma X11, forcing the window to be overlay fixed the issue:

![5urAnN.png](https://imgpile.com/images/5urAnN.png)

It works in Windowed mode but not Windowed Fullscreen, where APT shows as a black box in the middle of the screen.

**Comment by Gachick** [comment]
On Screen Display on KDE didn't work for me, but setting "Focus protection" property to "None" seemed to fix the issue.

**Comment by LuyeT** [comment]
Using i3/sway I always had an issue with the overlay soaking up the keyboard presses (though mouse input reached the game). For me the most robust solution was to just have poe-trade as a window on my secondary monitor, however, whenever the game loses focus (for example when checking trade in web browser), the window closes and i'll have to move the window back each time the game is refocused

`for_window [class="awakened-poe-trade"] move container $ws15
`
only works the first time poe trade window opens (i.e. first exec of poe-trade + first focus gain of PoE)

I'm wondering if it would be possible to have an "always open/windowed mode" option checkbox. I could maybe also hack this myself, though if you could point me in the right direction (source files relevant to overlay opening/closing), that would be greatly appreciated.

**Comment by SnosMe** [comment]
@LuyeT this mode is tracked in #803

**Comment by LuyeT** [comment]
thanks, don't know how i missed that one while searching for it

**Comment by liperium** [comment]
FOR KDE :
These are my settings, seems to work!
![image](https://user-images.githubusercontent.com/48138387/233846614-a95ab3c2-1027-4b2f-b8c0-47a1edca70e1.png)

**Comment by Xenorage** [comment]
Hey, I have the same problem In Arch/Gnome. When I do a price check a click  does not make the keyboard focus back to the game. Is there Window Rules on Gnome?

**Comment by ScuroGuardiano** [comment]
Hello there, I don't know if it's right place but I had a lot of troubles to make awakened-poe-trade work on awesomewm but after something like 10-12h of fighting I made it work flawlessly, so I am giving...
## The Solution
> Note: I play games in workspace with **floating** layout and with that it works for me ^^
### `picom.conf`
In picom if you have set `unredir-if-possible = true` you must exclude poe from it:
```
unredir-if-possible-exclude = ["class_g = 'awakened-poe-trade'", "class_g='steam_app_238960'"]
```
Unfortunately it can give some penalty to performance, if you don't want vsync you must disable it in picom and in game.
Also add APT to shadow-exclude or you can see some black little line on top of poe:
```
shadow-exclude = ["class_g = 'awakened-poe-trade'",];
```
### `awesome/rc.lua`
At the bottom of rules add this:
```lua
    {
        rule_any = {
            class = {
                "awakened-poe-trade"
            }
        },
        properties = {
            border_width = 0,
            titlebars_enabled = false,
            focus = false,
            ontop = true
        }
    }
```

Also add to function body `client.connect_signal("manage", function (c)` this:
```lua
    if c.class == "awakened-poe-trade" then
        gears.timer.delayed_call(function()
            c.ontop = true
        end)
    end
```
After that all shortcuts should work including CTRL+D but if that's not the case add also in `rc.lua`
```lua
client.connect_signal("property::ontop", function(c)
    if c.class == "awakened-poe-trade" and not c.ontop then
        gears.timer.delayed_call(function() c.ontop = true end)
    end
end)
```
And now it should work **almost** flawlessly, without needing to click 2x SHIFT+SPACE to make game working, CTRL+D working, it works as it works on Windows. Sometimes although you must click 2x SHIFT+SPACE to make CTRL+D working if you're switching focus but it happens rarely and it's hard to reproduce to further solve it.

[Here's my awesomewm and picom config for this fix](https://github.com/ScuroGuardiano/dotfiles/tree/446b74e76f72e0c8db4cfb3761e9074b2c4ab337/.config) (ignore i3, it's not working there correctly)

On other WM you must just force APT to be always on top and to not gain focus on spawn and propably make sure it's floating. I guess. And of course disable borders, titlebars and stuff.

## Explanation
APT worked very poorly for me, I had to slam SHIFT+SPACE 'til I could play normally. So I decided to notify every time new window is focused and I saw that it just looped "APT focus", "PoE focus", "APT focus", "PoE focus" etc.
Then with more digging I realized that when APT window is spawned PoE loses focus, in response APT window destroys itself coz PoE lost focus and APT wasn't activated by user, PoE gains focus, APT creates window and it loops. So we need to make sure that APT start unfocused, hence one of our rules:
```lua
focus = false
```
Unfortunately CTRL+D was still not working and I had to press 3x SHIFT+SPACE to make settings appear. I realized that the first time I clicked SHIFT+SPACE or CTRL+ALT+D on item APT window was spawned **below** Path of Exile, then second SHIFT+SPACE closed it and the next spawned it finally **above** Path of Exile. So I added this rule:
```lua
ontop = true
```
But it didn't work. Something like 2h of digging and I found out that yeah, when window is spawned it is set on top **BUT** after that APT disable ontop xD -> [here](https://github.com/SnosMe/electron-overlay-window/blob/28261ce92633292c9accd8e185174489311f0b1f/src/index.ts#L68) and [here](https://github.com/SnosMe/awakened-poe-trade/blob/eacdc4f890bd39b0a7463ca6d2129facbc530266/main/src/windowing/OverlayWindow.ts#L34)

I don't really know why [this line](https://github.com/SnosMe/electron-overlay-window/blob/28261ce92633292c9accd8e185174489311f0b1f/src/index.ts#L91) doesn't bring it to top, I don't know electron that much. Anyways with that knowledge the solution is simple - I just need to force APT to be always on top.
So I added delayed call to manage, so after APT window is created it just set ontop on true again. But I guess it's not working perfectly so I added also listener, if property ontop on APT window is changed to false, I just change it to true again.

And with that and of course disabling borders and titlebars, APT works *almost* ***perfectly*** on awesomewm. CTRL+D works, CTRL+ALT+D works, Shift+Space works UwU

PS. I love FOSS and ability to look at the source code of a program :heart:

**Comment by RyanHx** [comment]
Just to give an updated working configuration after experiencing this issue, running EndeavourOS (Arch), KDE Plasma (Wayland), Awakened Poe Trade version 3.25.104 (AppImage), using PoE in Windowed Fullscreen (borderless) mode:

![Image](https://github.com/user-attachments/assets/dff596d0-ba1b-4389-afe0-75f272a343d1)

Having `Accept Focus` forced to `No` as shown in @liperium's example meant the overlay could not be interacted with, even though it was brought to the front. Forcing it to the `On-screen display` layer seemed sufficient. `Skip taskbar` prevents the blinking taskbar icon from showing every time you use a hotkey like price checking.

I also set this up for the PoE window, running via Steam. If you're not using the Steam client you can use the **Detect Window Properties** button to populate your `Window class (application)`:

![Image](https://github.com/user-attachments/assets/479adb02-2a9d-4076-8aa0-00e037e52222)

Forcing `Focus stealing prevention` off, allowing other windows (the overlay) to take focus. Adaptive sync is unrelated.

**Comment by lordgreg** [comment]
> Hey, I have the same problem In Arch/Gnome. When I do a price check a click does not make the keyboard focus back to the game. Is there Window Rules on Gnome?

same here. current workaround:
- reopen awakened or
- price check, then immediatelly shift+space to show the main awakened window, price check window will also appear then

anyone knows for better solution in gnome?

**Comment by strokesws** [comment]
> Just to give an updated working configuration after experiencing this issue, running EndeavourOS (Arch), KDE Plasma (Wayland), Awakened Poe Trade version 3.25.104 (AppImage), using PoE in Windowed Fullscreen (borderless) mode:
> 
> ![Image](https://github.com/user-attachments/assets/dff596d0-ba1b-4389-afe0-75f272a343d1)
> 
> Having `Accept Focus` forced to `No` as shown in [@liperium](https://github.com/liperium)'s example meant the overlay could not be interacted with, even though it was brought to the front. Forcing it to the `On-screen display` layer seemed sufficient. `Skip taskbar` prevents the blinking taskbar icon from showing every time you use a hotkey like price checking.
> 
> I also set this up for the PoE window, running via Steam. If you're not using the Steam client you can use the **Detect Window Properties** button to populate your `Window class (application)`:
> 
> ![Image](https://github.com/user-attachments/assets/479adb02-2a9d-4076-8aa0-00e037e52222)
> 
> Forcing `Focus stealing prevention` off, allowing other windows (the overlay) to take focus. Adaptive sync is unrelated.

This should be added to the docs!

**Comment by kristiansja** [comment]
> Just to give an updated working configuration after experiencing this issue, running EndeavourOS (Arch), KDE Plasma (Wayland), Awakened Poe Trade version 3.25.104 (AppImage), using PoE in Windowed Fullscreen (borderless) mode:
> 
> ![Image](https://github.com/user-attachments/assets/dff596d0-ba1b-4389-afe0-75f272a343d1)
> 
> Having `Accept Focus` forced to `No` as shown in [@liperium](https://github.com/liperium)'s example meant the overlay could not be interacted with, even though it was brought to the front. Forcing it to the `On-screen display` layer seemed sufficient. `Skip taskbar` prevents the blinking taskbar icon from showing every time you use a hotkey like price checking.
> 
> I also set this up for the PoE window, running via Steam. If you're not using the Steam client you can use the **Detect Window Properties** button to populate your `Window class (application)`:
> 
> ![Image](https://github.com/user-attachments/assets/479adb02-2a9d-4076-8aa0-00e037e52222)
> 
> Forcing `Focus stealing prevention` off, allowing other windows (the overlay) to take focus. Adaptive sync is unrelated.

The settings work fine, but when I try to price something the translucent layer isnt there. And moving my mouse at all closes the trade window. I have tried most things, but nothing seems to work

<img width="2530" height="1413" alt="Image" src="https://github.com/user-attachments/assets/8d4dd1c3-5e0a-4cc9-8122-38897ed49600" />

**Comment by RyanHx** [comment]
@kristiansja I occasionally experience the same thing. Usually what worked for me was slowing down the key combination, like holding `Ctrl` for a second before pressing `D`, and maybe keeping `Ctrl` down for a second while moving the mouse over to the price check window. It's not an exact science as I never quite worked out what was fixing it specifically, but the window does end up staying open properly.

**Comment by kristiansja** [comment]
> [@kristiansja](https://github.com/kristiansja) I occasionally experience the same thing. Usually what worked for me was slowing down the key combination, like holding `Ctrl` for a second before pressing `D`, and maybe keeping `Ctrl` down for a second while moving the mouse over to the price check window. It's not an exact science as I never quite worked out what was fixing it specifically, but the window does end up staying open properly.

what the ****?! This works, but it seems completely fixable?

**Comment by andrewmkrause** [comment]
Just in case someone else runs into this, in addition to the window rules above (which you can access by hitting super button and typing "window rules") you also have to disable the option to confine mouse to window under UI within POE itself:

<img width="917" height="732" alt="Image" src="https://github.com/user-attachments/assets/4e45a844-eebd-417e-89fe-19abe5605305" />

This allows you to interact with the overlay after doing a price check on a rare (which doesn't search automatically)
